### PR TITLE
ssh/tailssh: support LDAP users for Tailscale SSH

### LINKED ISSUE #4945: ssh/tailssh: Cannot login as LDAP user on Linux host
### What is the issue?

I've set up Tailscale SSH on a Linux server where most of the users are from LDAP. For example:

```
# cat /etc/passwd | grep jsmith
# cat /etc/shadow | grep jsmith
# getent passwd | grep jsmith
jsmith:x:123456789:123456789:Smith, John:/home/jsmith:/bin/bash
```

When I try to SSH in as an LDAP user, I am denied with the following message:

```
$ ssh jsmith@myserver
failed to lookup jsmith
jsmith@myserver: Permission denied (tailscale).
```

I'm guessing that this corresponds to an error being returned by this call to `user.Lookup`:

https://github.com/tailscale/tailscale/blob/1d04e01d1e62e0f83a24c0e2963c1b74db33c00d/ssh/tailssh/tailssh.go#L208-L215

If I try to SSH in as a local (non-LDAP) user, it works successfully.

### Steps to reproduce

To reproduce this issue, first make sure that `LDAPUSER` is an LDAP user on your server. Then run:

```
ssh LDAPUSER@myserver
```

### Are there any recent changes that introduced the issue?

I recently did the following:
1. Updated Tailscale (by doing `sudo yum update tailscale`).
2. Ran `sudo tailscale up --ssh` to start the Tailscale SSH server.

### OS

Linux

### OS version

CentOS Linux release 7.9.2009 (Core)

### Tailscale version

1.26.1

### Bug report

BUG-53b053b3be60de1d1ace2edd177cb07c87b1f3b1a4c49b3b307d74b854fe7801-20220627044844Z-95747bf68628c167

### Other information

I installed Tailscale by doing adding the Tailscale repository and then doing `sudo yum install tailscale`.

```
# tailscale --version
1.26.1
  tailscale commit: b3687b70758d29f5fd6021c34d58eba7381812c0
  other commit: bcd70ce328082f464f89735d9adfa5bcd90ccf7e
  go version: go1.18.3-ts04d67b90d8
```

```
# cat /etc/system-release
CentOS Linux release 7.9.2009 (Core)
# uname -a
Linux myserver 3.10.0-1160.59.1.el7.x86_64 #1 SMP Wed Feb 23 16:47:03 UTC 2022 x86_64 x86_64 x86_64 GNU/Linux
```

### PR DESCRIPTION
**Pull request by bradfitz**
Fixes #4945

### CODE REVIEW

File: ssh/tailssh/incubator.go [hunk 1]
```diff
@@ -688,18 +688,11 @@ func (ss *sshSession) startWithStdPipes() (err error) {
 	return nil
 }
 
-func loginShell(u *user.User) string {
+func loginShell(u *userMeta) string {
+	if u.LoginShell != "" {
```
Comment (maisem) [inline ssh/tailssh/incubator.go:692]: add a comment that we expect this to be filled out on linux

**Review by maisem — APPROVED** [review]

File: ssh/tailssh/user.go [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+	for len(f) < 7 {
+		f = append(f, "")
+	}
+	um := &userMeta{
+		User: user.User{
+			Username: f[0],
+			Uid:      f[2],
+			Gid:      f[3],
+			Name:     f[4],
+			HomeDir:  f[5],
```
[the remark this answers is not in the document]
Comment (maisem) [reply ssh/tailssh/user.go:98]: by this i mean the `user.User` struct, which is really unlikely.

File: ssh/tailssh/user.go [hunk 3]
```diff
@@ -0,0 +1,117 @@
+// Copyright (c) Tailscale Inc & AUTHORS
+// SPDX-License-Identifier: BSD-3-Clause
+
+//go:build linux || (darwin && !ios) || freebsd || openbsd
+
+// Package tailssh is an SSH server integrated into Tailscale.
```
Comment (andrew-d) [inline ssh/tailssh/user.go:6]: Do you need the comment here?

**Review by andrew-d — APPROVED** [review]

**Comment by valscale** [comment]
I'm getting a test failure on MacOS that looks like it's caused by this commit:

Valeries-work-laptop:tailscale valerie$ git checkout 58ab66ec51f1963fbee302c75ad0017d81d37884^
HEAD is now at e8b06b22 version: properly handle vcs.modified when it's "false"
Valeries-work-laptop:tailscale valerie$ ./tool/go test ./ssh/tailssh/...
ok  	tailscale.com/ssh/tailssh	0.624s
Valeries-work-laptop:tailscale valerie$ git checkout 58ab66ec51f1963fbee302c75ad0017d81d37884
Previous HEAD position was e8b06b22 version: properly handle vcs.modified when it's "false"
HEAD is now at 58ab66ec ssh/tailssh: support LDAP users for Tailscale SSH
Valeries-work-laptop:tailscale valerie$ ./tool/go test ./ssh/tailssh/...
--- FAIL: TestSSHRecordingCancelsSessionsOnUploadFailure (0.00s)
    tailssh.go:238: ssh-conn-20230518T175857-383c8a0b74: handling conn: 100.100.100.101:2231->alice@100.100.100.102:22
    tailssh.go:238: ssh-conn-20230518T175857-383c8a0b74: failed to look up valerie: strconv.Atoi: parsing "valerie": invalid syntax
    --- FAIL: TestSSHRecordingCancelsSessionsOnUploadFailure/upload-denied (0.00s)
        tailssh_test.go:416: client: ssh: handshake failed: ssh: unable to authenticate, attempted methods [none], no supported methods remain
    tailssh.go:238: ssh-conn-20230518T175857-c54a06d660: handling conn: 100.100.100.101:2231->alice@100.100.100.102:22
    tailssh.go:238: ssh-conn-20230518T175857-c54a06d660: failed to look up valerie: strconv.Atoi: parsing "valerie": invalid syntax
    --- FAIL: TestSSHRecordingCancelsSessionsOnUploadFailure/upload-fails-after-starting (0.00s)
        tailssh_test.go:416: client: ssh: handshake failed: ssh: unable to authenticate, attempted methods [none], no supported methods remain
--- FAIL: TestMultipleRecorders (1.00s)
    tailssh.go:238: ssh-conn-20230518T175857-94e7830942: handling conn: 100.100.100.101:2231->alice@100.100.100.102:22
    tailssh.go:238: ssh-conn-20230518T175857-94e7830942: failed to look up valerie: strconv.Atoi: parsing "valerie": invalid syntax
    tailssh_test.go:518: client: ssh: handshake failed: ssh: unable to authenticate, attempted methods [none], no supported methods remain
    tailssh_test.go:545: timed out waiting for recording
--- FAIL: TestSSHRecordingNonInteractive (1.00s)
    tailssh_test.go:607: client: ssh: handshake failed: ssh: unable to authenticate, attempted methods [none], no supported methods remain
    tailssh_test.go:632: EOF
--- FAIL: TestSSHAuthFlow (0.01s)
    --- FAIL: TestSSHAuthFlow/accept (0.00s)
        tailssh_test.go:777: banner = "Welcome to Tailscale SSH!"
        tailssh_test.go:773: unexpected banner: "failed to look up valerie\r\n"
        tailssh_test.go:790: client: ssh: handshake failed: ssh: unable to authenticate, attempted methods [none], no supported methods remain
    --- FAIL: TestSSHAuthFlow/simple-check (0.00s)
        tailssh_test.go:775: banner = "failed to look up valerie\r\n"; want "Welcome to Tailscale SSH!"
        tailssh_test.go:790: client: ssh: handshake failed: ssh: unable to authenticate, attempted methods [none], no supported methods remain
        tailssh_test.go:816: missing banners: [Welcome to Tailscale SSH!]
    --- FAIL: TestSSHAuthFlow/multi-check (0.00s)
        tailssh_test.go:777: banner = "First"
        tailssh_test.go:775: banner = "failed to look up valerie\r\n"; want "url-here"
        tailssh_test.go:790: client: ssh: handshake failed: ssh: unable to authenticate, attempted methods [none], no supported methods remain
        tailssh_test.go:816: missing banners: [url-here Welcome to Tailscale SSH!]
    --- FAIL: TestSSHAuthFlow/check-reject (0.00s)
        tailssh_test.go:777: banner = "First"
        tailssh_test.go:775: banner = "failed to look up valerie\r\n"; want "Go Away!"
        tailssh_test.go:816: missing banners: [Go Away!]
    --- FAIL: TestSSHAuthFlow/force-password-auth (0.00s)
        tailssh_test.go:777: banner = "Welcome to Tailscale SSH!"
        tailssh_test.go:773: unexpected banner: "failed to look up valerie\r\n"
        tailssh_test.go:790: client: ssh: handshake failed: ssh: unable to authenticate, attempted methods [none], no supported methods remain
FAIL
FAIL	tailscale.com/ssh/tailssh	2.421s
FAIL

**Pull request merged**